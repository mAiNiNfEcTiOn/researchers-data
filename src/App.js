import axios from 'axios';
import React, { useEffect, useState, Suspense } from 'react';
import { createBrowserHistory } from 'history';
import { Router, Switch, Route } from 'react-router-dom';

import Navigation from './components/Navigation/Navigation';
import ResearcherData from './components/ResearcherData/ResearcherData';

import styles from './App.module.css';
import researchersDataNormalizer from './helpers/researchersDataNormalizer';

const history = createBrowserHistory();

function App() {
  const [loading, setLoading] = useState(true);
  const [data, setData] = useState(null);
  const [error, setError] = useState(null);

  useEffect(() => {
    setLoading(true);

    if (process.env.REACT_APP_API_URL) {
      const source = axios.CancelToken.source();

      axios(process.env.REACT_APP_API_URL, { cancelToken: source.token })
        .then((response) => {
          setData(researchersDataNormalizer(response.data));
          setLoading(false);
        })
        .catch((e) => {
          if (!axios.isCancel(e)) {
            setError(e);
            setLoading(false);
          }
        });

      return () => {
        source.cancel('Operation canceled by the user.');
      };
    }

    import('./mockedData.json')
      .then((response) => {
        setData(researchersDataNormalizer(response.default));
        setLoading(false);
      });
  }, []);

  if (loading) {
    return null; // TODO: Show a loading placeholder
  }

  if (error) {
    return <div>An error has occurred!</div>; // TODO: Handle errors;
  }

  return (
    <Suspense fallback={null}>
      <Router history={history}>
        <header className={styles.header}>
          <h1>Prominent Researchers</h1>
        </header>
        <main className={styles.main}>
          <Switch>
            <Route exact path="/" render={() => <Navigation researchers={data} />} />
            <Route
              exact
              path="/researcher/:id"
              render={({ match }) => {
                const researcherId = parseInt(match.params.id, 10);
                return (
                  <>
                    <Navigation researchers={data} selectedResearcherId={researcherId} />
                    <ResearcherData researcher={data.find((researcher) => researcher.id === researcherId)} />
                  </>
                );
              }}
            />
          </Switch>
        </main>
      </Router>
    </Suspense>
  );
}

export default App;
