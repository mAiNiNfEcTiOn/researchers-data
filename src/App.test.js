import axios from 'axios';
import React from 'react';
import { render, waitForElement, fireEvent, wait, queryByText } from '@testing-library/react';
import { Line } from 'react-chartjs-2';

import App from './App';
import mockedData from './mockedData.json'
import styles from './App.module.css';
import navigationStyles from './components/Navigation/Navigation.module.css';
import researcherDataStyles from './components/ResearcherData/ResearcherData.module.css';

jest.mock('axios');
jest.mock('react-chartjs-2', () => ({
  Line: jest.fn(),
}));

describe('App', () => {
  afterEach(() => {
    delete process.env.REACT_APP_API_URL;
  });

  it('renders default layout without the chart', async () => {
    const { container } = render(<App />);

    const main = await waitForElement(() => container.querySelector(`.${styles.main}`));
    const researchers = main.querySelectorAll(`.${navigationStyles.researcherCardLink}`);
    const chart = main.querySelector(`.${researcherDataStyles.researcherData}`);

    expect(container).toMatchSnapshot();
    expect(researchers.length).toBe(5);
    expect(chart).toBeFalsy();
  });

  it('when selecting the a researcher it displays his/her data in the chart', async () => {
    let receivedDataset;
    Line.mockImplementation(({ data }) => {
      receivedDataset = data;
      return <canvas />;
    });
    const { container } = render(<App />);

    const main = await waitForElement(() => container.querySelector(`.${styles.main}`));
    const researchers = main.querySelectorAll(`.${navigationStyles.researcherCardLink}`);

    fireEvent.click(researchers[2]);
    await wait(() => researchers[2].classList.contains(navigationStyles.researcherCardSelected));

    expect(container).toMatchSnapshot();
    expect(receivedDataset).toMatchObject({
      labels: [2016, 2017, 2018, 2019, 2020],
      datasets: [
        {
          borderColor: '#A7761C',
          data: [199, 301, 198, 176, 252],
          fill: false,
          label: 'Citations',
          lineTension: 0,
          pointBackgroundColor: '#FFF',
          pointBorderWidth: 3,
          pointHoverBackgroundColor: '#FFF',
          pointHoverRadius: 7,
          pointRadius: 5,
          spanGaps: true,
        },
        {
          borderColor: '#666666',
          data: [77, 103, 47, 81, 85],
          fill: false,
          label: 'Publications',
          lineTension: 0,
          pointBackgroundColor: '#FFF',
          pointBorderWidth: 3,
          pointHoverBackgroundColor: '#FFF',
          pointHoverRadius: 7,
          pointRadius: 5,
        },
      ],
    });
  });

  it('uses axios if the REACT_APP_API_URL is set', async () => {
    process.env.REACT_APP_API_URL = 'https://fake.api.dom';
    axios.CancelToken = { source: () => ({ token: 'fake token', cancel: jest.fn() }) }
    axios.mockResolvedValueOnce({Â data: mockedData });

    const { container } = render(<App />);

    await waitForElement(() => container.querySelector(`.${styles.main}`));

    expect(container).toMatchSnapshot();
  });

  it('renders empty when axios returns an error', async () => {
    process.env.REACT_APP_API_URL = 'https://fake.api.dom';
    axios.CancelToken = { source: () => ({ token: 'fake token', cancel: jest.fn() }) }
    axios.mockRejectedValueOnce(new Error('Oops error... not a cancellation'));

    const { container } = render(<App />);

    await waitForElement(() => queryByText(container, 'An error has occurred!'));

    expect(container).toMatchSnapshot();
  });
});
