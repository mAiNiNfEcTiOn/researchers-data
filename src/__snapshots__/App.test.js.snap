// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders default layout without the chart 1`] = `
<div>
  <header
    class="header"
  >
    <h1>
      Prominent Researchers
    </h1>
  </header>
  <main
    class="main"
  >
    <section
      class="researchers"
    >
      <nav>
        <ul>
          <li
            class="researcherCard"
          >
            <a
              class="researcherCardLink"
              href="/researcher/10005"
            >
              <div
                class="researcherCardAvatar"
              />
              <div
                class="researcherCardData"
              >
                <h2>
                  John Doe
                </h2>
                <p
                  class="affiliation"
                >
                  California Institute of Technology
                </p>
                <p
                  class="stats"
                >
                  493
                   Outputs · 
                  1385
                   Citations
                </p>
              </div>
            </a>
          </li>
          <li
            class="researcherCard"
          >
            <a
              class="researcherCardLink"
              href="/researcher/10007"
            >
              <div
                class="researcherCardAvatar"
              />
              <div
                class="researcherCardData"
              >
                <h2>
                  Joe Bloggs
                </h2>
                <p
                  class="affiliation"
                >
                  Harvard University
                </p>
                <p
                  class="stats"
                >
                  443
                   Outputs · 
                  1228
                   Citations
                </p>
              </div>
            </a>
          </li>
          <li
            class="researcherCard"
          >
            <a
              class="researcherCardLink"
              href="/researcher/10035"
            >
              <div
                class="researcherCardAvatar"
              />
              <div
                class="researcherCardData"
              >
                <h2>
                  Jane Doe
                </h2>
                <p
                  class="affiliation"
                >
                  Massachusetts Institute of Technology
                </p>
                <p
                  class="stats"
                >
                  393
                   Outputs · 
                  1126
                   Citations
                </p>
              </div>
            </a>
          </li>
          <li
            class="researcherCard"
          >
            <a
              class="researcherCardLink"
              href="/researcher/10012"
            >
              <div
                class="researcherCardAvatar"
              />
              <div
                class="researcherCardData"
              >
                <h2>
                  Tom Smith
                </h2>
                <p
                  class="affiliation"
                >
                  Stanford University
                </p>
                <p
                  class="stats"
                >
                  368
                   Outputs · 
                  1092
                   Citations
                </p>
              </div>
            </a>
          </li>
          <li
            class="researcherCard"
          >
            <a
              class="researcherCardLink"
              href="/researcher/10023"
            >
              <div
                class="researcherCardAvatar"
              />
              <div
                class="researcherCardData"
              >
                <h2>
                  Harry Stone
                </h2>
                <p
                  class="affiliation"
                >
                  University of Oxford
                </p>
                <p
                  class="stats"
                >
                  343
                   Outputs · 
                  1032
                   Citations
                </p>
              </div>
            </a>
          </li>
        </ul>
      </nav>
    </section>
  </main>
</div>
`;

exports[`App renders empty when axios returns an error 1`] = `
<div>
  <div>
    An error has occurred!
  </div>
</div>
`;

exports[`App uses axios if the REACT_APP_API_URL is set 1`] = `
<div>
  <header
    class="header"
  >
    <h1>
      Prominent Researchers
    </h1>
  </header>
  <main
    class="main"
  >
    <section
      class="researchers"
    >
      <nav>
        <ul>
          <li
            class="researcherCard"
          >
            <a
              class="researcherCardLink"
              href="/researcher/10005"
            >
              <div
                class="researcherCardAvatar"
              />
              <div
                class="researcherCardData"
              >
                <h2>
                  John Doe
                </h2>
                <p
                  class="affiliation"
                >
                  California Institute of Technology
                </p>
                <p
                  class="stats"
                >
                  493
                   Outputs · 
                  1385
                   Citations
                </p>
              </div>
            </a>
          </li>
          <li
            class="researcherCard"
          >
            <a
              class="researcherCardLink"
              href="/researcher/10007"
            >
              <div
                class="researcherCardAvatar"
              />
              <div
                class="researcherCardData"
              >
                <h2>
                  Joe Bloggs
                </h2>
                <p
                  class="affiliation"
                >
                  Harvard University
                </p>
                <p
                  class="stats"
                >
                  443
                   Outputs · 
                  1228
                   Citations
                </p>
              </div>
            </a>
          </li>
          <li
            class="researcherCard"
          >
            <a
              class="researcherCardLink researcherCardSelected"
              href="/"
            >
              <div
                class="researcherCardAvatar"
              />
              <div
                class="researcherCardData"
              >
                <h2>
                  Jane Doe
                </h2>
                <p
                  class="affiliation"
                >
                  Massachusetts Institute of Technology
                </p>
                <p
                  class="stats"
                >
                  393
                   Outputs · 
                  1126
                   Citations
                </p>
              </div>
            </a>
          </li>
          <li
            class="researcherCard"
          >
            <a
              class="researcherCardLink"
              href="/researcher/10012"
            >
              <div
                class="researcherCardAvatar"
              />
              <div
                class="researcherCardData"
              >
                <h2>
                  Tom Smith
                </h2>
                <p
                  class="affiliation"
                >
                  Stanford University
                </p>
                <p
                  class="stats"
                >
                  368
                   Outputs · 
                  1092
                   Citations
                </p>
              </div>
            </a>
          </li>
          <li
            class="researcherCard"
          >
            <a
              class="researcherCardLink"
              href="/researcher/10023"
            >
              <div
                class="researcherCardAvatar"
              />
              <div
                class="researcherCardData"
              >
                <h2>
                  Harry Stone
                </h2>
                <p
                  class="affiliation"
                >
                  University of Oxford
                </p>
                <p
                  class="stats"
                >
                  343
                   Outputs · 
                  1032
                   Citations
                </p>
              </div>
            </a>
          </li>
        </ul>
      </nav>
    </section>
    <section
      class="researcherData"
    >
      <canvas />
    </section>
  </main>
</div>
`;

exports[`App when selecting the a researcher it displays his/her data in the chart 1`] = `
<div>
  <header
    class="header"
  >
    <h1>
      Prominent Researchers
    </h1>
  </header>
  <main
    class="main"
  >
    <section
      class="researchers"
    >
      <nav>
        <ul>
          <li
            class="researcherCard"
          >
            <a
              class="researcherCardLink"
              href="/researcher/10005"
            >
              <div
                class="researcherCardAvatar"
              />
              <div
                class="researcherCardData"
              >
                <h2>
                  John Doe
                </h2>
                <p
                  class="affiliation"
                >
                  California Institute of Technology
                </p>
                <p
                  class="stats"
                >
                  493
                   Outputs · 
                  1385
                   Citations
                </p>
              </div>
            </a>
          </li>
          <li
            class="researcherCard"
          >
            <a
              class="researcherCardLink"
              href="/researcher/10007"
            >
              <div
                class="researcherCardAvatar"
              />
              <div
                class="researcherCardData"
              >
                <h2>
                  Joe Bloggs
                </h2>
                <p
                  class="affiliation"
                >
                  Harvard University
                </p>
                <p
                  class="stats"
                >
                  443
                   Outputs · 
                  1228
                   Citations
                </p>
              </div>
            </a>
          </li>
          <li
            class="researcherCard"
          >
            <a
              class="researcherCardLink researcherCardSelected"
              href="/"
            >
              <div
                class="researcherCardAvatar"
              />
              <div
                class="researcherCardData"
              >
                <h2>
                  Jane Doe
                </h2>
                <p
                  class="affiliation"
                >
                  Massachusetts Institute of Technology
                </p>
                <p
                  class="stats"
                >
                  393
                   Outputs · 
                  1126
                   Citations
                </p>
              </div>
            </a>
          </li>
          <li
            class="researcherCard"
          >
            <a
              class="researcherCardLink"
              href="/researcher/10012"
            >
              <div
                class="researcherCardAvatar"
              />
              <div
                class="researcherCardData"
              >
                <h2>
                  Tom Smith
                </h2>
                <p
                  class="affiliation"
                >
                  Stanford University
                </p>
                <p
                  class="stats"
                >
                  368
                   Outputs · 
                  1092
                   Citations
                </p>
              </div>
            </a>
          </li>
          <li
            class="researcherCard"
          >
            <a
              class="researcherCardLink"
              href="/researcher/10023"
            >
              <div
                class="researcherCardAvatar"
              />
              <div
                class="researcherCardData"
              >
                <h2>
                  Harry Stone
                </h2>
                <p
                  class="affiliation"
                >
                  University of Oxford
                </p>
                <p
                  class="stats"
                >
                  343
                   Outputs · 
                  1032
                   Citations
                </p>
              </div>
            </a>
          </li>
        </ul>
      </nav>
    </section>
    <section
      class="researcherData"
    >
      <canvas />
    </section>
  </main>
</div>
`;
